### ISSUE TITLE
Is RexRay actively maintained?

### ISSUE THREAD

**Issue by jocatalin** [rexray/rexray#1338]
# Summary
As I don't see any activity since January 2019 on this repository, I'm wandering if the project is still maintained. My use case is the integration with open-source[ DC/OS external persistent storage](https://docs.d2iq.com/mesosphere/dcos/1.12/storage/external-storage/).
Will new features be added?
Will new bugs be fixed? 

The question is mainly for the the vmware Rex-Ray project maintainers: @codenrhoden @akutz @clintkitson

**Comment by clintkitson** [comment]
Hello Catalin,

Have you looked at CSI? CSI was introduced a couple of years ago and is the
focus of numerous storage providers for interop with K8s/Mesos. Generally,
community PRs are welcome across the repo and releases are as needed at
this point. Our focus is on parts of REX-Ray such as the `gocsi` library.

**Comment by a-nldisr** [comment]
Looking at https://github.com/rexray/rexray/issues/1337 you could probably draw your conclusion there.

**Comment by kallisti5** [comment]
Sigh.  Literally went live with this solution last December for a large open source project.  I ended up scaling it back due to #1331  glad I did :-|

Anyone interested in forking / maintaining or taking over the project from the original author? (if you can find them)

**Comment by clintkitson** [comment]
Sorry for this @kallisti5. If there are community members that are interested in maintaining please do let me know.

**Comment by jaschaio** [comment]
@kallisti5 what did you use instead?

@clintkitson I would love to help but don't even know what language rexray is written in :/ Looks like you are the last "active" maintainer of this repo, is that correct?

**Comment by Trojan295** [comment]
@clintkitson I'm willing to help maintain this project. Kinda sucks most of the CSI available implementations are done mostly for k8s and e.g. in case of AWS ECS we have to use RexRay for ECS <-> EBS integration, so RexRay is great here.

**Comment by clintkitson** [comment]
Glad to hear you're making use of it @Trojan295. Do you have any thoughts about the direction you would want to take the project?

**Comment by wolph** [comment]
If the rexray maintainers are willing to give up control that would be useful, otherwise a fork wouldn't be a bad idea. It's a shame how projects like these get abandoned... probably just the company shifting priorities but still... at the very least hand it over to someone that cares

**Comment by Enquier** [comment]
@clintkitson Given the lack of clear direction from Mirantis and the closed sourcing/deletion of CloudStor documentation I'd really like to see this pick up as it is the only remaining solution for Docker Swarm (especially for AWS and OpenStack). 
If there are some people interested i'd love to see a group start picking this back up. Some ideas i've had:
* Cleanup and Close existing PR's
* Stand up community resources
* Work to implement/refactor both Rexray and it's documentation to highlight it's use as a "native" CSI solution/bridge for Swarm

I have my own fork that brings in several PR's for and a few of my own fixes for the OpenStack Cinder but I'd be happy to work with a few folks bring them and many of the others in. I've noticed that the community (especially the Swarm community) is starting to just take for granted that we are "second-string" despite the fact that most of us are firm believers that it is a better solution (for us anyway) than K8s, mesos etc. I have experienced what Mirantis support looks like, (at least in the short term) and if we want to save Swarm we are going to have to do it ourselves.

/rallycry

**Comment by iFloris** [comment]
@Enquier yes please. We make use of docker in swarm mode extensively using [Caprover](https://github.com/caprover/caprover) and have now reached the point where we really need swarm mode data persistence, which we planned to achieve using Rexray/s3fs.

**Comment by clintkitson** [comment]
I am open to having a community Zoom call to further discuss. Please thumbs
up if you would attend and reply if there are any special considerations
for time etc.

**Comment by cPanelRicky** [comment]
@Enquier @clintkitson I am willing to throw my hat in to help with keeping this project going.

I'm actually using the Hashicorp Nomad Orchestrator on Openstack and this projects Cinder Docker Volume Plugin makes life so much easier. 

Yeah, Nomad is slowly adding CSI support but, even though it's supposed to be a "platform agnostic" specification, I have yet to come across a single CSI driver that documents its deployment process. 

The docker volume plugins however are truly orchestrator agnostic and it took me about 3 minutes to find the information I needed to install/configure the plugin and have migratable persistant storage across my cluster.

@Enquier Can you share what fixes you made in the cinder driver (as mentioned before, this is what i'm using Rexray for)? I started digging into the Cinder driver to look into adding Application Credentials support but was curious if that may have been one of the things you worked on in your own fork?

**Comment by Enquier** [comment]
@cPanelRicky Check-out logitainer/rexray I wasn't sure if @clintkitson was more interested in handing off or starting a fork so I moved mine over to that as a starting point in the event of option #2. I'm in the process of updating the whole stack to use Go Mod as opposed to Dep so let me know if you encounter any issues (see logitainer/csi-nfs,csi-blockdevice,csi-vfs etc)

**Comment by RickyGrassmuck** [comment]
@Enquier Perfect, appreciate it!

I started attempting to migrate over to Go modules as well but found that there are going to be some modules that have become incompatible over the last year or so (mainly the Gophercloud module) so that's gonna take some work to get fixed up.

**Comment by Enquier** [comment]
Just following up, busy summer. Anyone have a feeling one-way or the other?

**Comment by saulmmendoza** [comment]
@Enquier  Yoo hello, Im interested, Im building a PoC using Docker Swarm and I was searching for a persistent storage solution until the algorithm showed me this repository

where can I found a list of issues we should solve? how should we prioritize them?  
Because I read that some people are using OpenStack, others DigitalOcean, and I was looking to use another provider which is not listed (I would write the code for it so no worries), but would rather write that later, I think there are more important issues like the guy hitting the api limits on DigitalOcean (which probably will apply to another cloud providers)

**Comment by jostmart** [comment]
So @Enquier how did this turn out? I'm looking at logitainer/rexray and it look stalled :(
I'm looking for a good storage solution for Swarm. My platform is Vmware Simplivity and I can't see any support for it.

**Comment by acouvreur** [comment]
I have the same issue @jostmart. Running a swarm cluster with nodes on VMWare platform. We sadly can't use the VSFD plugin : https://github.com/vmware-archive/vsphere-storage-for-docker.

I'm in need of a storage provider that can use vSphere as a provider. 

Has anyone tried to create a vSphere provider for REX-Ray ?

**Comment by noelmcloughlin** [comment]
Apparently, VMware maintains this.  

https://blogs.vmware.com/cloudnative/2018/06/22/rex-ray-persists-your-questions-answered/

> Q. Is REX-Ray still being maintained?  A: Yes, by VMware.
> Q. To what extent is VMware maintaining REX-Ray? A: VMware has committed multiple open-source engineers to REX-Ray.
> Q. I’ve been a part of the REX-Ray community for a long time, where should I go for continued support and collaboration?
> A: The REX-Ray GitHub repository.

**Comment by iFloris** [comment]
> Apparently, VMware maintains this.  
> 
> 
> 
> https://blogs.vmware.com/cloudnative/2018/06/22/rex-ray-persists-your-questions-answered/
> 
> 
> 
> > Q. Is REX-Ray still being maintained?  A: Yes, by VMware.
> 
> > Q. To what extent is VMware maintaining REX-Ray? A: VMware has committed multiple open-source engineers to REX-Ray.
> 
> > Q. I’ve been a part of the REX-Ray community for a long time, where should I go for continued support and collaboration?
> 
> > A: The REX-Ray GitHub repository.

That's quite unexpected. But cool!

**Comment by Enquier** [comment]
Note that blog post is nearly two years old. I’m fairly certain that VMWare
has since created a separate CSI based plugin elsewhere.

**Comment by Enquier** [comment]
> So @Enquier how did this turn out? I'm looking at logitainer/rexray and it look stalled :(
> I'm looking for a good storage solution for Swarm. My platform is Vmware Simplivity and I can't see any support for it.

It’s still there. It’s still my intention that it be maintained but I was looking for a signal one way or the other from the original team.

 It seems to me the most valuable piece here is maintaining the CSI compatible interface for swarm and that should allow other CSI plug-ins for K8s PV to work for swarm. So factoring that out and then identifying working solutions from other sources seems to be the right approach

**Comment by noelmcloughlin** [comment]
> I’m fairly certain that VMWare has since created a separate CSI based plugin elsewhere.

Yes but VMWare gave a commitment to the Open Source community. Therefore they are obliged to ensure the project is maintained (nobody is asking for new features, just bugs fixed, maintenance releases and commitment adhered to).

**Comment by Enquier** [comment]
I’d appreciate any support from VMware but at this point all of us using non K8 solutions are on our own. 

So, I’ve expressed willingness to start working the crank. I have a solution that works for me over at logitainer/rexray and other than working to add native efs support I don’t have a clear feature path. 

So at this point we have two options, either fracture out and start pushing CSI at our individual orchestration platform level at which point we’d still want to maintain the CSI features that are here but nowhere else (Ie cinder and S3) or continue on to rexray as a platform for CSI. I think the best solution is a hybrid so, I’d  like to start by getting some of the PR’s closed out. So I would put out a call to anyone who made a PR to clean them up and reprise them in Logitainer and we can start to triage them. We will need to sort out the build pipelines from there

**Comment by Galbar** [comment]
Hello, I saddened to see the status of this project 😢 .

I come here because I just ran into an issue with the dobs integration. I looked at the code and opened a PR only to 5 minutes later realize that someone else had opened another PR for the same issue with a very similar solution. Would it be possible to merge it? https://github.com/rexray/rexray/pull/1344

Also, I've tried building the project locally but I'm unable to do it. Help?

**Comment by a-nldisr** [comment]
> > I’m fairly certain that VMWare has since created a separate CSI based plugin elsewhere.
> 
> Yes but VMWare gave a commitment to the Open Source community. Therefore they are obliged to ensure the project is maintained (nobody is asking for new features, just bugs fixed, maintenance releases and commitment adhered to).

No, they are not obliged to do anything. Making software Opensource does not mean there is a contract to maintain or even pay attention to your issues, you are however free to ignore certain contributors or companies in the future. Depending on the Opensource license you are however free to fork the project and run it as your own. 

It would be really polite of them to warn others that this project is stale and no longer developed, but again there is no rule that they should. It is up to you to judge them and consider alternatives if you come across a solution of theirs. Not much else you can do.

**Comment by ryanmasse** [comment]
@clintkitson the website and installer are offline as bintray.com is EOL.  Any chance this could be resolved?

**Comment by josephlim75** [comment]
is it possible to build and host the binaries in github releases instead of bintray ?  Since rexray provided instruction to build the go binaries, that isn't difficult to do.  I believe many people are getting Forbidden when they reach out to bintray.com
I am using rexray for docker swarm in GCP.  I'm just trying out.  The issue I'm facing is during resizing of volume disk size, it doesn't seem to be working correctly

**Comment by griffmick** [comment]
I've had very poor experiences with Rexray in the past and do not recommend it.  Rclone offers a much better (maintained!) volume driver for Docker Swarm: https://rclone.org/docker/

Have been using this volume driver in production with no issues for well over a year.

**Comment by DeanWronowski** [comment]
Is there an alternative to using Rexray?

**Comment by iFloris** [comment]
> Is there an alternative to using Rexray?

We moved to https://github.com/sapk/docker-volume-rclone first, then later migrated our entire Docker Swarm setup to https://github.com/rancher/rancher with https://github.com/longhorn/longhorn, but there is an rclone-CSI as well https://github.com/wunderio/csi-rclone. 

But as @griffmick said, rclone has a native volume driver, which is probably the best directly comparable alternative to rexray. https://github.com/rexray/rexray/issues/1338#issuecomment-1338987270

**Comment by cballesteros-sistrol** [comment]
I have only started to have this problem today in one of my environments.... Am I the only one?

**Comment by Ten0** [comment]
any of these support EBS? (https://github.com/rexray/rexray/issues/1282#issuecomment-1600973555)

**Comment by bram-tv** [comment]
Looking at [dockerhub ](https://hub.docker.com/search?q=rexray&sort=updated_at&order=desc) it appears that @nbryant42 fork is the one with the most recent changes.

@nbryant42 any chance you can clarify the status of your fork? (It might also help to enable the "Issues"-feature on your fork - which can be done in the settings so that issues - if any - can be created on it)
(Also not clear if the runlo user on dockerhub is you or someone else publishing images of your clone)

**Comment by nbryant42** [comment]
@bram-tv,

My fork was pretty recently updated for newer toolchain and library dependencies based on govulncheck (purely a compliance measure - I doubt anything was ever exploitable in something like this, but we like to play it safe). I'd have to check, but this was just a couple weeks ago so it should still be clean per govulncheck.

However I've just recently moved on from the employer where that work was performed and I've not currently got access to any AWS test systems. I'd be willing to keep merging toolchain and dependency updates (this is only a few minutes of work every few weeks) but will not be able to test them unless things change. It might be time to see if anyone else is interested in keeping this project fresh.

Also, I don't know who runlo is. Looks like somebody building their own image from my fork.

**Comment by bram-tv** [comment]
> ... I've not currently got access to any AWS test systems. I'd be willing to keep merging toolchain and dependency updates (this is only a few minutes of work every few weeks) but will not be able to test them unless things change.

I am able to assist with the testing; at least when it's clear what should be tested.
I only recently began using rexray so not entirely sure what should/shouldn't be tested.

> It might be time to see if anyone else is interested in keeping this project fresh.

It's unfortunate that rexray still seems to be the "best" way to link ECS with a (statefull) EBS...
(Please: someone proof me wrong, I would be happy to switch to something else but so far there appears to be nothing else)

**Comment by nbryant42** [comment]
@bram-tv, my process had been:

- Commit dependency and toolchain updates as needed (plus the occasional random bugfix)
- Push that code to my active branch (`revert-1345`) in the fork
- Build EBS driver image and push to the `latest` tag in the repo I was managing
- If this seems to work fine (as in, it's mounting volumes successfully--just basic testing really), re-push to the `stable` tag
